### PR TITLE
OTAManager: Limit number of saved update packages

### PR DESCRIPTION
**Pull request by zwim**
Over the time there are a lot of downloaded update packages on the device.

Added a menu entry to set the maximum number of saved download packages and a method to clean the download directory.

Only tested on the emulator and on an Android mobile, so someone should test it on other devices too.

grafik

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8538)

### CODE REVIEW

**Comment by poire-z** [comment]
Any reason not to hardcode a limit of 3 or 5 - or even 1 ? (So we don't need to add a menu item.)

**Comment by Frenzie** [comment]
> Only tested on the emulator and on an Android mobile, so someone should test it on other devices too.

Other devices don't work that way, so there's nothing to test. ;-) The way it works on devices using zsync2 is:

1. Here's the current package
2. Here's the package we might want to update to
3. zsync2 cleverly deduces which few hundred kB to download
4. New package is built
5. (On the next start) Old files overwritten with new files.

> Any reason not to hardcode a limit of 3 or 5 - or even 1 ? (So we don't need to add a menu item.)

Rollbacks, potentially? Also I'd say it's rather unclear if users will appreciate files being deleted from under their nose in their downloads folder. I certainly wouldn't expect it.

**Comment by Frenzie** [comment]
There will be three or four files.

* koreader/ota/koreader-kobo-latest-nightly.zsync
* koreader/ota/koreader-kobo-latest-stable.zsync
* koreader.installed.tar
* package.index

None of these should be deleted.

File: frontend/ui/otamanager.lua [hunk 1]
```diff
@@ -335,6 +337,8 @@ function OTAManager:fetchAndProcessUpdate()
                 end
             })
         end
+
+        self:cleanUpdates()
```
Comment (Frenzie) [inline frontend/ui/otamanager.lua:341]:
```suggestion
        if self:getOTAType() == "link" then
            self:cleanUpdates()
        end
```

**Comment by Galunid** [comment]
> Rollbacks, potentially? Also I'd say it's rather unclear if users will appreciate files being deleted from under their nose in their downloads folder. I certainly wouldn't expect it.

To be fair, keeping 2-3 most recent updates should be more than enough. The way I see it, I see it more like cleaning the mess after ourselves, so that user doesn't have to bother.

**Comment by Frenzie** [comment]
> To be fair, keeping 2-3 most recent updates should be more than enough. The way I see it, it's more like cleaning the mess after ourselves, so that user doesn't have to bother.

This feels like a concern most likely to affect… well, us, but perhaps ironically we might like to keep a few old builds around.

As such, I'd actually propose something similar to the script @NiLuJe and I wrote over on the koreader.rocks server, where we make a distinction between stable and nightly, and keep… well, I forget the exact numbers but a number of stables and a number of nightlies. (There's no big clever rationale behind it, other than disk space limits.)

http://build.koreader.rocks/download/stable/
http://build.koreader.rocks/download/nightly/

File: frontend/ui/otamanager.lua [hunk 2]
```diff
@@ -481,10 +485,65 @@ function OTAManager:getOTAMenuTable()
...
         }
     }
 end
 
+function OTAManager:cleanUpdates()
+    if self.nb_ota_packages <= 0 then
+        return
+    end
+
+    local update_template = "^koreader-" .. self:getOTAModel() .. "-.*-v.*"
```
Comment (Frenzie) [inline frontend/ui/otamanager.lua:523]: Regarding my comment about distinguishing stables and nightlies, if Version:getNormalizedCurrentVersion() returns a commit hash it's not a stable. Anyway, just a thought.

**Comment by NiLuJe** [comment]
That's only a concern on Android, though, right?

**Comment by NiLuJe** [comment]
(in which case, possibly simply remove the apk on first launch after a successful update?).

**Comment by zwim** [comment]
Some thoughts:
We can set the number of backup updates to zero (then everything gets deleted) and use a negative number for unlimited.
And set the default to zero (for normal users).

We can move the setting to developer options.

(We can make two counters, one for nightlies and one for stable, also I won't like that.)

**Comment by pazos** [comment]
> That's only a concern on Android, though, right?

I think so.

> simply remove the apk on first launch after a successful update?

Now we have a FileProvider, so we can probably store APKs in `/sdcard/android/data/org.koreader.launcher/files/cache` and attempt to wipe said dir after every succesful app update (ie: after a mismatch between bundled vs installed assets). 

I'm not sure if the bundled download manager is able to download to private app dirs or we need to implement our own. Implementing our own is a drag but makes sense, specially for Tolinos, where the downloadManager is utterly broken.

**Comment by zwim** [comment]
> > simply remove the apk on first launch after a successful update?
> 
> Now we have a FileProvider, so we can probably store APKs in `/sdcard/android/data/org.koreader.launcher/files/cache` and attempt to wipe said dir after every succesful app update (ie: after a mismatch between bundled vs installed assets).

Please note that this cache has to be in a subfolder of "/files/cache" otherwise there is a problem (this was a problem during implementation of coverimage.koplugin)

> 
> I'm not sure if the bundled download manager is able to download to private app dirs or we need to implement our own. Implementing our own is a drag but makes sense, specially for Tolinos, where the downloadManager is utterly broken.

Does that mean, I should stop working on this PR now, or will we use it (with a different path to the new cache)?

**Comment by zwim** [comment]
In the latest commit, the menu entry is moved to the developer-menu (so well hidden), the default is set to zero, so no cache.
Write-protected files don't get deleted.
Not ready to merge, as I want to wait for tomorrows nightly to check ther "zero" function on my phone.

**Comment by Frenzie** [comment]
> Does that mean, I should stop working on this PR now, or will we use it (with a different path to the new cache)?

I don't think doing it from here rather than Kotlin would make much sense?

**Comment by zwim** [comment]
Ok, then I will stop here. I leave this open until the kotlin implementation is ready.
BTW: The zero updates option does not work in the current commit (on Android ), but I have a ready solution for that.

**Comment by zwim** [comment]
@pasos: Do you plan to work on implementing this functionality in Kotlin? Of should we stay with the proposed PR?

**Comment by pazos** [comment]
> Do you plan to work on implementing this functionality in Kotlin? Of should we stay with the proposed PR?

@zwim: I won't have time until the summer if we need to wrap up our own download async task. But I don't like this PR as it adds new options for no reason :/.

I wouldn't mind one of the following options, in order of preference:

1. Remove apks  after asset extraction, after https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/Assets.kt#L28
2. this PR without user options :)

The benefit of 1 vs 2 is that we make sure the app was updated before prunning files.

The main benefit of reimplementing the wheel (aka DownloadManager) is getting something that works on Tolinos. (I understand that each time you're updating you get a prompt saying **can't download, retry with the browser?**). Then the browser cannot store stuff in our private dirs.

If tolinos were not a special snowflake I would suggest to use [setDestinationInExternalFilesDir](https://developer.android.com/reference/android/app/DownloadManager.Request#setDestinationInExternalFilesDir(android.content.Context,%20java.lang.String,%20java.lang.String)) insteade of `setDestinationInExternalPublicDir` in https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/ApkUpdater.kt#L36

**Comment by zwim** [comment]
In the last commit, only the update package belonging to the currently running KOReader version gets deleted.
(If you want to keep the update package, you can create the file `koreader\keep_updates`). 

>I won't have time until the summer if we need to wrap up our own download async task. But I don't like this PR as it adds new options for no reason :/.

This is meant as a quick solution until summer. Once the "async download task" is ready, this PR can be reverted.

**Comment by zwim** [comment]
>The main benefit of reimplementing the wheel (aka DownloadManager) is getting something that works on Tolinos. (I understand that each time you're updating you get a prompt saying can't download, retry with the browser?). Then the browser cannot store stuff in our private dirs.

@pazos You understand this right, but in my case I have wiped the `Tolino.apk` completely and with that even the browser has gone. So I have to sideload the apk anyway.

**Comment by pazos** [comment]
> You understand this right, but in my case I have wiped the `Tolino.apk` completely and with that even the browser has gone. So I have to sideload the apk anyway.

In that case there're no APKs on your device and you're fine :)

> This is meant as a quick solution until summer. Once the "async download task" is ready, this PR can be reverted

Still, all of this is android code, much better handled in luajit-launcher.

**Comment by pazos** [comment]
My proposal is in https://github.com/koreader/android-luajit-launcher/pull/346

For the 99% of people that would be an improvement. We no longer clutter the Downloads folder of the user. Instead we store apks in /sdcard/Android/org.koreader.launcher/files and prune them after they were correctly installed.

Tolino users won't get that change since the download is not handled by our app but by the browser, so they need to treat our apks like any other downloaded content.  That's the bit that can be changed if we implement our own download service.

**Comment by zwim** [comment]
Close this in favor of https://github.com/koreader/android-luajit-launcher/pull/346

**Pull request closed without merging**